### ISSUE TITLE
SC2187 - Ash scripts will be checked as Dash, unsolvable.

### ISSUE THREAD

**Issue by frenck** [hadolint/hadolint#234]
- [x] This is a bug report
- [ ] This is a feature request
- [ ] I searched existing issues before opening this one

Running Hadolint on an Alpine based Dockerfile, it will be impossible to fix the SC2187 shellcheck warning.

`SC2187 Ash scripts will be checked as Dash. Add '# shellcheck shell=dash' to silence.`

### Expected behavior

Detect we are using `ash` or offer a way to suppress the message as described in:

<https://github.com/koalaman/shellcheck/wiki/SC2187>

### Actual behavior

The message pops up (listed in OP), and I see no way to fix this.

### Steps to reproduce the behavior

**Output of `hadolint --version` or
  `docker run --rm hadolint/hadolint hadolint --version`:**

```bash
Haskell Dockerfile Linter v1.10.0-0-g3abbe56
```

**Dockerfile** (if relevant)

```Dockerfile
# shellcheck shell=dash
FROM alpine:3.8

# Set shell
SHELL ["/bin/ash", "-o", "pipefail", "-c"]

# Install base system
RUN apk add --no-cache curl=7.60.0-r1
```

PS: Removing the `# shellcheck shell=dash` line, or moving it to other parts of the file, doesn't make a difference.

**Additional environment details (OS, stack version, etc.)**

Tested on MacOS High Sierra, Ubuntu Bionic and using your official Docker container (all latest available versions).

**Comment by frenck** [comment]
Extra info: The 1.9.0 does not have this issue.
The warning is only suppressible using `.hadolint.yaml`

**Comment by lorenzo** [comment]
I’ll make this a warning if it’s by default, it is not very useful

**Comment by frenck** [comment]
I do not totally agree with the suppression.
I really want to be able to specify the shell script type used.

Take for example this Dockerfile:

```
FROM hassioaddons/base:2.0.0

# Install base system
RUN apk add --no-cache curl=7.60.0-r1
```

So which shell is it going to use with shellcheck?

**Comment by lorenzo** [comment]
> So which shell is it going to use with shellcheck?

In that example `sh` is detected. Since there is no previous `SHELL` declaration before that run instruction. Currently the only way (and probably also the only safe way) to hint what the sell type is by using the `SHELL` instruction

In the case of detecting ash, it will be treated as `dash` there is not workaround for that. Either we try to be helpful and treated as `dash` or we don't do any analysis at all.

**Comment by frenck** [comment]
Hmm... I was unclear I guess.

I was suggesting adding:

`# hadolint shell=dash`

This was, I'm still able to specify the shell, with the `SHELL` command (which would introduce an extra layer, just for the sake of linting)

**Comment by lorenzo** [comment]
`SHELL` does not introduces extra layers, though. Only the instructions `RUN`, `COPY`, `ADD` create layers.

I think I did understand you, my concern with `# hadolint shell=dash` is that it may get out of sync with the parent image, whereas specifying `SHELL` again in the child image cannot. That's what I want with `SHELL` being the safe option

**Comment by lorenzo** [comment]
@frenck can you open a new ticket with that suggestion, please?

I think this particular issue is solved, as it was a regression introduced after adding the automatic shell type detection (before the bash syntax was used regardless) and I think your suggestion has merit and Id like a separate issue to track it.

**Comment by frenck** [comment]
Thank you for the explanation. Agreed.

In that case, we should add a rule that warns when SHELL is not specified.
